keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,607 @@
+
+- Alpha: Feature flag for new controller is disabled by default
+- Dual controller. Both old and new implementation co-exist
+
+#### Release 1.21
+
+- Beta: Feature flag for new controller is enabled by default. If the distribution chooses it can be disabled
+- Introduce CronJob in batch/v1
+
+#### Release 1.22
+
+- GA: The feature flag is deprecated and the old controller code is cleaned up
+
+## Graduation Criteria
+
+- [ ] Implement shared informers to reduce pressure on API Server
+- [ ] Pass conformance tests
+- [ ] Update documents reflecting the changes
+- [ ] Pass CronJob e2e tests
+- [ ] Pass CronJob unit-tests
+- [ ] Pass scale tests
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: `CronJobControllerV2`
+    - Components depending on the feature gate: `kube-controller-manager`
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  No changes are expected.
+
+* **Are there any tests for feature enablement/disablement?**
+  Not applicable, given we're missing framework allowing switching feature-gates during e2e.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  This might fail when leader election for kube-controller-manager ends up assigned to two separate instances. In this case the two controller might end up fighting over the cronjob resource.  Additionally, a bug in the new controller might trigger a panic and thus crashing the whole kube-controller-manager.  Another problem might come from GA-ing batch/v1.CronJob which might cause problems with conversions between different versions of that resource.
+
+* **What specific metrics should inform a rollback?**
+  Unexpected restarts of kube-controller-manager.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  Manual test is planned once the implementation is finished.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  batch/v1beta1.CronJob is already deprecated.
+
+### Monitoring Requirements
+
+* **How can an operator determine if the feature is in use by workloads?**
+  For the new controller part checking `cronjob_job_creation_skew` and `cronjob_controller_rate_limiter_use` metrics, accompanied with a set of queue-related metrics the old controller is not using.  For the CronJob resource GA verifying if CronJob exists in `batch/v1` API group in the cluster.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [x] Metrics
```

Comment:
I would suggest listing the total jobs present as well.  If memory serves, we've had some issues in the past.

Reply:
It would have to be some kind of jobs per cronjob metric, b/c we don't want to account jobs not created by cronjob controller.

Reply:
I've added `etcd_object_counts{resource="jobs.batch"}` to the metrics.